pkg/http/server.go

```diff
@@ -0,0 +1,195 @@
+	inShutdown             bool                                // indicates whether the server is in shutdown or not
+	requestCount           int32                               // counter holds no. of request in process.
+}
+
+// Start - start HTTP server
+func (srv *Server) Start() (err error) {
+	// Take a copy of server fields.
+	tlsConfig := srv.TLSConfig
+	readTimeout := srv.ReadTimeout
+	writeTimeout := srv.WriteTimeout
+	handler := srv.Handler
+
+	addrs := srv.Addrs
+	tcpKeepAliveTimeout := srv.TCPKeepAliveTimeout
+	updateBytesReadFunc := srv.UpdateBytesReadFunc
+	updateBytesWrittenFunc := srv.UpdateBytesWrittenFunc
+	errorLogFunc := srv.ErrorLogFunc
+
+	// Create new HTTP listener.
+	var listener *httpListener
+	listener, err = newHTTPListener(
+		addrs,
+		tlsConfig,
+		tcpKeepAliveTimeout,
+		readTimeout,
+		writeTimeout,
+		updateBytesReadFunc,
+		updateBytesWrittenFunc,
+		errorLogFunc,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Wrap given handler to do additional
+	// * automatic redirect to HTTPS if given TLS is configured.
+	// * return 503 (service unavailable) if the server in shutdown.
+	wrappedHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt32(&srv.requestCount, 1)
+		defer atomic.AddInt32(&srv.requestCount, -1)
+
+		// If TLS is configured and the request is sent as non-TLS, return 307 (temporary redirect).
+		if tlsConfig != nil && r.TLS == nil {
+			u := *(r.URL)
+			u.Scheme = "https"
+			u.Host = r.Host
+			http.Redirect(w, r, u.String(), http.StatusTemporaryRedirect)
+			return
+		}
+
+		// If server is in shutdown, return 503 (service unavailable)
+		srv.inShutdownMutex.Lock()
+		inShutdown := srv.inShutdown
+		srv.inShutdownMutex.Unlock()
+		if inShutdown {
+			w.WriteHeader(http.StatusServiceUnavailable)
+			return
+		}
+
+		// Handle request using passed handler.
+		handler.ServeHTTP(w, r)
+	})
+
+	srv.listenerMutex.Lock()
+	srv.Handler = wrappedHandler
+	srv.listener = listener
+	srv.listenerMutex.Unlock()
+
+	// Start servicing with listener.
+	return srv.Server.Serve(listener)
+}
+
+// Shutdown - shuts down HTTP server.
+func (srv *Server) Shutdown() error {
+	srv.inShutdownMutex.Lock()
+	if !srv.inShutdown {
+		// Indicate the server is in shutdown.
+		srv.inShutdown = true
+		srv.inShutdownMutex.Unlock()
+
+		// Close underneath HTTP listener.
+		srv.listenerMutex.Lock()
+		err := srv.listener.Close()
+		srv.listenerMutex.Unlock()
+
+		// Wait for opened connection to be closed up to Shutdown timeout.
+		shutdownTimeout := srv.ShutdownTimeout
+		shutdownTimeoutCh := time.After(shutdownTimeout)
+		ticker := time.NewTicker(serverShutdownPoll)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-shutdownTimeoutCh:
+				return errors.New("timed out. some connections are still active. doing abnormal shutdown")
+			case <-ticker.C:
+				if atomic.LoadInt32(&srv.requestCount) <= 0 {
```

Comment:
There a problem in this which is exposed in this PR and fixed https://github.com/minio/minio/pull/4500#issuecomment-306980377 

requestCount if at all is `1` then this loop will never honor `shutdownTimeoutCh` since `time.After()` only gets invoked if the other case is waiting for 5secs > . 

Basically this would wait for ConnTimeout i.e 30secs until when requestCount will be `1` for Go rpc connections. So gracefulShutdownTimeout is never invoked. Just like how it is in current master.

Can you check @balamurugana